File: src/lib/logger.ts
const PROJECT = 'GAUSHALA';
type LogLevel = 'ERROR' | 'WARN' | 'INFO' | 'DEBUG';

function formatEntry(level: LogLevel, layer: string, file: string, message: string, context?: Record<string, unknown>): string {
  const timestamp = new Date().toISOString().replace('T', ' ').split('.')[0];
  const ctx = context ? ' | ' + Object.entries(context).map(([k, v]) => `${k}=${JSON.stringify(v)}`).join(' ') : '';
  return `[${timestamp}] [${level}] [${PROJECT}] [${layer.toUpperCase()}] [${file}] ${message}${ctx}`;
}

function write(level: LogLevel, layer: string, file: string, message: string, context?: Record<string, unknown>) {
  const entry = formatEntry(level, layer, file, message, context);
  if (level === 'ERROR') console.error(entry);
  else if (level === 'WARN') console.warn(entry);
  else console.log(entry);
}

const logger = {
  error: (layer: string, file: string, message: string, context?: Record<string, unknown>) =>
    write('ERROR', layer, file, message, context),
  warn: (layer: string, file: string, message: string, context?: Record<string, unknown>) =>
    write('WARN', layer, file, message, context),
  info: (layer: string, file: string, message: string, context?: Record<string, unknown>) =>
    write('INFO', layer, file, message, context),
  debug: (layer: string, file: string, message: string, context?: Record<string, unknown>) =>
    write('DEBUG', layer, file, message, context),
};

export default logger;

File: pipeline/lib/logger.py
import os
import json
from datetime import datetime

LOG_DIR = '/home/openclaw/logs/gaushala-network'
PROJECT = 'GAUSHALA'

def ensure_log_dir():
    os.makedirs(LOG_DIR, exist_ok=True)

def get_log_file(layer: str) -> str:
    date = datetime.now().strftime('%Y-%m-%d')
    return os.path.join(LOG_DIR, f'{layer}-{date}.log')

def write(level: str, layer: str, file: str, message: str, context: dict = None):
    try:
        ensure_log_dir()
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        ctx = ''
        if context:
            ctx = ' | ' + ' '.join(f'{k}={json.dumps(v)}' for k, v in context.items())
        entry = f'[{timestamp}] [{level}] [{PROJECT}] [{layer.upper()}] [{file}] {message}{ctx}\n'
        print(entry.strip())
        with open(get_log_file(layer), 'a') as f:
            f.write(entry)
        if level == 'ERROR':
            with open(get_log_file('errors'), 'a') as f:
                f.write(entry)
    except Exception as e:
        print(f'Logger failed: {e}')

class Logger:
    def __init__(self, layer: str, file: str):
        self.layer = layer
        self.file = file

    def error(self, message: str, context: dict = None):
        write('ERROR', self.layer, self.file, message, context)

    def warn(self, message: str, context: dict = None):
        write('WARN', self.layer, self.file, message, context)

    def info(self, message: str, context: dict = None):
        write('INFO', self.layer, self.file, message, context)

    def debug(self, message: str, context: dict = None):
        write('DEBUG', self.layer, self.file, message, context)
